### PR TITLE
BLS12-377 curve operations

### PR DESCRIPTION
**Pull request by shamatar**
This EIP adds a set of precompiles to efficiently implement operations over BLS12-377 curve

### CODE REVIEW

File: EIPS/eip-2539.md [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+x is negative = false
+```
+
+#### Fine points and encoding of base elements
+
+##### Field elements encoding:
+
+To encode points involved in the operation one has to encode elements of the base field and the extension field.
+
+Base field element (Fp) is encoded as `48` bytes by performing BigEndian encoding of the corresponding (unsigned) integer. Corresponding integer **must** be less than field modulus.
```
Comment (iAmMichaelConnor) [inline EIPS/eip-2539.md:81]: @shamatar to be compatible with 32-byte 'words' of the EVM/Solidity (and to allow easy mload and mstore operations of 32-byte values in assembly), perhaps elements of Fp should be encoded as 64 bytes instead?

The first 16-bytes could be enforced to be all zeros, or ignored by the builtin contracts.

It'd be great to get your thoughts.

File: EIPS/eip-2539.md [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+For elements of the quadratic extension field (Fp2) encoding is byte concatenation of individual encoding of the coefficients totaling in `96` bytes for a total encoding. For an Fp2 element in a form `el = c0 + c1 * v` where `v` is formal quadratic non-residue and `c0` and `c1` are Fp elements the corresponding byte encoding will be `encode(c0) || encode(c1)` where `||` means byte concatenation.
```
Comment (iAmMichaelConnor) [inline EIPS/eip-2539.md:83]: @shamatar perhaps elements of Fp2 could be 128 bytes, to enable easier field arithmetic in solidity (since real and 'imaginary' coefficients could each be stored in a uint[2] or bytes32[2]).

Comment (shamatar) [reply EIPS/eip-2539.md:83]: I’ll make similar adjustments to 2537 and will largely copy-paste and updated spec here

**Comment by prestwich** [comment]
@shamatar We are considering this EIP for inclusion in [Celo Blockchain](https://github.com/celo-org/celo-blockchain). Has thought been given to updating the precompile list with map to group operations in order to match the #2537 API?

**Comment by shamatar** [comment]
Sure, but I'm not aware of mapping parameters for BLS12-377. The easiest way is to copy-paste #2537 text in here while keeping the curve parameters

**Comment by prestwich** [comment]
I don't know that it's strictly necessary for anyone's use case. But it would be nice to have direct feature parity with the other precompile

File: EIPS/eip-2539.md [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+discussions-to: https://ethereum-magicians.org/t/eip-2539-bls12-377-precompile-discussion-thread/4659
+status: Draft
+type: Standards Track
+category: Core
+created: 2020-02-26
+requires: 1109, 2046
+---
+
+## Simple Summary
+This precompile adds operation on BLS12-377 curve (from Zexe paper) as a precompile in a set necessary to *efficiently* perform operations such as BLS signature verification and perform SNARKs verifications. Unique properties of BLS12-377 also later allow to have SNARKs that check BLS12-377 pairing in an efficient way and allow e.g. constant-size BLS signature aggregation.
```
Comment (MicahZoltu) [inline EIPS/eip-2539.md:14]: Remove "from Zexe paper".  This is a technical document, not a place to give props/kudos/attribution.  If you think that the paper is particularly relevant, I recommend adding it to `../assets/eip-2539/zexe-paper.pdf` and linking to it in a place where it makes sense that a future reader of this specification would care.

Comment (MicahZoltu) [reply EIPS/eip-2539.md:14]:
```suggestion
Adds BLS12-377 curve operations as precompiles.
```
Simple summary should be a very terse one-line description.  Longer than the title, but not meant to convey much information and instead let the reader quickly identify "am I in the right place?".  This may show up in a sub-heading somewhere for example.  The rest of this should probably go in the motivation, as it is is discussing *why* this is useful/interesting.

File: EIPS/eip-2539.md [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+- BLS12_377_G1MUL - to perform point multiplication on a curve defined over prime field
+- BLS12_377_G1MULTIEXP - to perform multiexponentiation on a curve defined over prime field
+- BLS12_377_G2ADD - to perform point addition on a curve twist defined over quadratic extension of the base field
+- BLS12_377_G2MUL - to perform point multiplication on a curve twist defined over quadratic extension of the base field
+- BLS12_377_G2MULTIEXP - to perform multiexponentiation on a curve twist defined over quadratic extension of the base field
+- BLS12_377_PAIRING - to perform a pairing operations between a set of *pairs* of (G1, G2) points
+
+Multiexponentiation operation is included to efficiently aggregate public keys or individual signer's signatures during BLS signature verification.
+
+### Proposed addresses table
```
Comment (MicahZoltu) [inline EIPS/eip-2539.md:30]:
```suggestion
### Addresses Table
```
EIPs don't propose, they assert.

File: EIPS/eip-2539.md [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+|BLS12_377_G1ADD          | 0x13  |
+|BLS12_377_G1MUL          | 0x14  |
+|BLS12_377_G1MULTIEXP     | 0x15  |
+|BLS12_377_G2ADD          | 0x16  |
+|BLS12_377_G2MUL          | 0x17  |
+|BLS12_377_G2MULTIEXP     | 0x18  |
+|BLS12_377_PAIRING        | 0x19  |
+
+## Motivation
+Motivation of this precompile is to add a cryptographic primitive that allows to get 120+ bits of security for operations over pairing friendly curve compared to the existing BN254 precompile that only provides 80 bits of security. In addition it allows efficient one-time recursive proof aggregations, e.g. proofs about existence of BLS12-377 based signature.
```
Comment (MicahZoltu) [inline EIPS/eip-2539.md:43]: Consider adding some of the stuff from Simple Summary to here.  This is a great place to tell users *why* this EIP is a useful standard.

File: EIPS/eip-2539.md [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+X c0 = 0x018480be71c785fec89630a2a3841d01c565f071203e50317ea501f557db6b9b71889f52bb53540274e3e48f7c005196
+X c1 = 0x00ea6040e700403170dc5a51b1b140d5532777ee6651cecbe7223ece0799c9de5cf89984bff76fe6b26bfefa6ea16afe
+Y c0 = 0x00690d665d446f7bd960736bcbb2efb4de03ed7274b49a58e458c282f832d204f2cf88886d8c7c2ef094094409fd4ddf
+Y c1 = 0x00f8169fd28355189e549da3151a70aa61ef11ac3d591bf12463b01acee304c24279b83f5e52270bd9a1cdd185eb8f93
+Pairing parameters:
+|x| (miller loop scalar) = 0x8508c00000000001
+x is negative = false
+```
+
+#### Fine points and encoding of base elements
```
Comment (MicahZoltu) [inline EIPS/eip-2539.md:79]:
```suggestion
### Fine points and encoding of base elements
```
You skipped `###` here, this can sometimes break some rendering systems particularly when they try to render table of contents.

Comment (MicahZoltu) [reply EIPS/eip-2539.md:79]: Same thing with all of the following sections, they are all 1 too deep.

File: EIPS/eip-2539.md [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+|BLS12_377_PAIRING        | 0x19  |
+
+## Motivation
+Motivation of this precompile is to add a cryptographic primitive that allows to get 120+ bits of security for operations over pairing friendly curve compared to the existing BN254 precompile that only provides 80 bits of security. In addition it allows efficient one-time recursive proof aggregations, e.g. proofs about existence of BLS12-377 based signature.
+
+## Specification
+
+Curve parameters:
+
+BLS12-377 curve is fully defined by the following set of parameters (coefficient `A=0` for all BLS12 curves):
```
Comment (MicahZoltu) [inline EIPS/eip-2539.md:49]:
```suggestion
[BLS12-377](../assets/eip-2539/bls12-377.pdf) curve is fully defined by the following set of parameters (coefficient `A=0` for all BLS12 curves):
```
This would be a good spot to have a relative link out to an attachment of the BLS12-377 paper. (note: if you accept this comment, make sure to actually add the paper as an asset!)

File: EIPS/eip-2539.md [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+Motivation section covers a total motivation to have operations over BLS12-377 curve available. We also extend a rationale for move specific fine points.
+
+#### Multiexponentiation as a separate call
+
+Explicit separate multiexponentiation operation that allows one to save execution time (so gas) by both the algorithm used (namely Peppinger algorithm) and (usually forgotten) by the fact that `CALL` operation in Ethereum is expensive (at the time of writing), so one would have to pay non-negigible overhead if e.g. for multiexponentiation of `100` points would have to call the multipication precompile `100` times and addition for `99` times (roughly `138600` would be saved).
+
+## Backwards Compatibility
+There are no backward compatibility questions.
+
+## Important notes
```
Comment (MicahZoltu) [inline EIPS/eip-2539.md:226]: This is a non-standard section, and while non-standard sections aren't technically disallowed we prefer they not be included so the repository remains consistent.  Can we move this into a subsection of the specification?  It *feels* pretty specification like to me, though I suppose it is really just a "pro-tip".

File: EIPS/eip-2539.md [hunk 9]
```diff
@@ -0,0 +1,266 @@
+---
+eip: 2539
+title: BLS12-377 curve operations
+author: Alex Vlasov (@shamatar)
+discussions-to: https://ethereum-magicians.org/t/eip-2539-bls12-377-precompile-discussion-thread/4659
+status: Draft
+type: Standards Track
+category: Core
+created: 2020-02-26
+requires: 1109, 2046
```
Comment (MicahZoltu) [inline EIPS/eip-2539.md:10]:
```suggestion
```
Why are these required?  It feels like a chain who doesn't have these implemented would still be able to include this EIP.

**Review by MicahZoltu — APPROVED** [review]
Looks good!  I left some feedback that I encourage checking out but nothing that blocks merging as a draft.

**Comment by rdubois-crypto** [comment]
Sidenote: the divisibility of (p-1) by 2^44 in BLS12-377 Fp field could also benefit to Starks (PQ-snarks), as the FRI requires NTT operations to interpolate the trace polynomial. https://vitalik.ca/general/2017/11/22/starks_part_2.html.

**Comment by burdges** [comment]
It's not discussed here, but..

IRTF hash-to-curve spec says Edwards curves should use Elligator for their hash-to-curve, not WB or SWU.  BLS12-377 should fall under this, given it has an Edwards form (and an extremely prominent one).
